src/api/cvc4cpp.h

```diff
@@ -0,0 +1,2357 @@
+     */
+    std::string toString() const;
+  private:
+    /* The internal expression wrapped by this operator term. */
+    CVC4::Expr* d_expr;
+};
+
+/**
+ * Serialize an operator term to given stream.
+ * @param out the output stream
+ * @param t the operator term to be serialized to the given output stream
+ * @return the output stream
+ */
+std::ostream& operator<< (std::ostream& out, const OpTerm& t) CVC4_PUBLIC;
+
+/**
+ * Hash function for OpTerms.
+ */
+struct CVC4_PUBLIC OpTermHashFunction
+{
+  size_t operator()(const OpTerm& t) const;
+};
+
+
+/* -------------------------------------------------------------------------- */
+/* Datatypes                                                                  */
+/* -------------------------------------------------------------------------- */
+
+class DatatypeConstructorIterator;
+class DatatypeIterator;
+
+/**
+ * A place-holder sort to allow a DatatypeDecl to refer to itself.
+ * Self-sorted fields of DatatypeDecls will be properly sorted when a Sort is
+ * created for the DatatypeDecl.
+ */
+class CVC4_PUBLIC DatatypeDeclSelfSort
+{
+};
+
+/**
+ * A CVC4 datatype selector declaration.
+ */
+class CVC4_PUBLIC DatatypeSelectorDecl
+{
+  friend class DatatypeConstructorDecl;
+  public:
+    /**
+     * Constructor.
+     * @param name the name of the datatype selector
+     * @param sort the sort of the datatype selector
+     * @return the DatatypeSelectorDecl
+     */
+    DatatypeSelectorDecl(const std::string& name, Sort sort);
+
+    /**
+     * Constructor.
+     * @param name the name of the datatype selector
+     * @param sort the sort of the datatype selector
+     * @return the DAtatypeSelectorDecl
+     */
+    DatatypeSelectorDecl(const std::string& name, DatatypeDeclSelfSort sort);
+
+    /**
+     * @return a string representation of this datatype selector
+     */
+    std::string toString() const;
+
+  private:
+    /* The name of the datatype selector. */
+    std::string d_name;
+    /* The sort of the datatype selector. */
+    Sort d_sort;
+};
+
+/**
+ * A CVC4 datatype constructor declaration.
+ */
+class CVC4_PUBLIC DatatypeConstructorDecl
+{
+  friend class DatatypeDecl;
+  public:
+    /**
+     * Constructor.
+     * @param name the name of the datatype constructor
+     * @return the DatatypeConstructorDecl
+     */
+    DatatypeConstructorDecl(const std::string& name);
+
+    /**
+     * Destructor.
+     */
+    ~DatatypeConstructorDecl();
+
+    /**
+     * Add datatype selector declaration.
+     * @param stor the datatype selector declaration to add
+     */
+    void addSelector(const DatatypeSelectorDecl& stor);
+
+    /**
+     * @return a string representation of this datatype constructor declaration
+     */
+    std::string toString() const;
+
+  private:
+    /* The internal (intermediate) datatype constructor wrapped by this
+     * datatype constructor declaration. */
+    std::shared_ptr<CVC4::DatatypeConstructor> d_ctor;
```

Comment:
Why is this a shared_ptr instead of a unique_ptr? Here is a general guideline that suggests preferring unique_ptr to shared_ptr by default: https://herbsutter.com/2013/05/29/gotw-89-solution-smart-pointers/ . I mostly just want this to be clear. Consider documenting this decision in the code somewhere.

Reply:
Because DatatypeConstructor (the internal one) do not have reference counting. Same for all other Datatype* types.

Reply:
Makes sense. Can this be documented somewhere?

Reply:
Yes, will do when i address your review.

Reply:
Done.